app/services/proforma_service/convert_proforma_task_to_task.rb

```diff
@@ -52,15 +60,24 @@ def files
     end
 
     def file_from_proforma_file(proforma_task_file)
-      task_file = TaskFile.new(file_attributes(proforma_task_file))
+      task_file = (@task.id && TaskFile.all_files_for_task(@task).find_by(xml_id: proforma_task_file.id)) || TaskFile.new
```

Comment:
Wouldn't it be more efficient to query all files once and then cache the result?

What do you think about the following implementation?

```rb
  # models/task_file.rb

  validate :unique_xml_id, if: -> { !fileable.nil? && xml_id_changed? }

  def unique_xml_id
    task = fileable.is_a?(Task) ? fileable : fileable.task
    xml_ids = (task.all_files(cached: false) - [self]).map(&:xml_id)
    errors.add(:xml_id, :not_unique) if xml_ids.include? xml_id
  end
```

```rb
  # models/task.rb
  def all_files(cached: true)
    return @all_files if cached && defined?(@all_files)

    @all_files = (files + tests.map(&:files) + model_solutions.map(&:files)).flatten
  end
```

```suggestion
      task_file = @task&.all_files&.select{|file| file.xml_id == proforma_task_file.id}&.first || TaskFile.new
```

Then, the new scope wouldn't be needed any longer and the tests you've written could be moved to `Task#all_files` (where we don't have any tests yet).

Reply:
that seems easier :+1: